xarray/core/utils.py

```diff
@@ -266,7 +292,10 @@ def dict_equiv(first, second, compat=equivalent):
     return True
 
 
-def ordered_dict_intersection(first_dict, second_dict, compat=equivalent):
+def ordered_dict_intersection(first_dict: Mapping[K, V],
+                              second_dict: Mapping[K, V],
+                              compat: Callable[[V, V], bool] = equivalent
+                              ) -> MutableMapping[K, V]:
```

Comment:
This should probably be `'OrderedDict[K, V]'`

Reply:
That's painful to achieve, because typing.OrderedDict was added in python 3.7.2, and I'm scared of hacking about what is such a fundamental class for xarray.
Also, are there any methods that OrderedDict offers that MutableMapping doesn't?

Reply:
If you put it in quotes, it's only checked when running mypy. That seems to work without any hacking, e.g., see 
https://github.com/pydata/xarray/blob/742ed3984f437982057fd46ecfb0bce214563cb8/xarray/core/merge.py#L28

Reply:
that would mean that anybody with less than Python 3.7.2 will get errors when running mypy...

Reply:
Personally, I'm OK with that. But it may be worth clarifying some cohesive policy.